catkit2/services/hamamatsu_camera/hamamatsu_camera.py

```diff
@@ -0,0 +1,494 @@
+            else:
+                self.make_property(name, lambda: getattr(self, name), lambda val: setattr(self, name, val))
+
+        make_property_helper('exposure_time')
+        make_property_helper('gain')
+        make_property_helper('brightness')
+
+        make_property_helper('width')
+        make_property_helper('height')
+        make_property_helper('offset_x')
+        make_property_helper('offset_y')
+
+        make_property_helper('sensor_width', read_only=True)
+        make_property_helper('sensor_height', read_only=True)
+
+        self.make_command('start_acquisition', self.start_acquisition)
+        self.make_command('end_acquisition', self.end_acquisition)
+
+        self.temperature_thread = threading.Thread(target=self.monitor_temperature)
+        self.temperature_thread.start()
+
+    def main(self):
+        """
+        The main function of the service.
+
+        This function is called when the service is started.
+        It starts the acquisition loop and waits for incoming frames.
+        """
+        while not self.should_shut_down:
+            if self.should_be_acquiring.wait(0.05):
+                self.acquisition_loop()
+
+    def close(self):
+        self.cam.dev_close()
+        self.cam = None
+
+        dcam.Dcamapi.uninit()
+
+    def acquisition_loop(self):
+        """
+        The main acquisition loop.
+
+        This function is the main loop for acquiring images from the camera.
+        It starts the acquisition and then waits for incoming frames.
+        """
+        # Make sure the data stream has the right size and datatype.
+        has_correct_parameters = np.allclose(self.images.shape, [self.height, self.width])
+        if not has_correct_parameters:
+            self.images.update_parameters('float32', [self.height, self.width], self.NUM_FRAMES)
+
+        # Start acquisition.
+        if self.cam.buf_alloc(self.NUM_FRAMES) is False:
+            raise RuntimeError(f'Dcam.buf_alloc() fails with error {self.cam.lasterr()}')
+        if self.cam.cap_start() is False:
+            raise RuntimeError(f'Dcam.cap_start() fails with error {self.cam.lasterr()}')
+        self.is_acquiring.submit_data(np.array([1], dtype='int8'))
+
+        timeout_millisec = 2000
+        try:
+            while self.should_be_acquiring.is_set() and not self.should_shut_down:
+                if self.cam.lasterr().is_timeout():
+                    self.log.warning('Timeout while waiting for frame')
+                if self.cam.wait_capevent_frameready(timeout_millisec) is False:
+                    raise RuntimeError(f'Dcam.wait_capevent_frameready({timeout_millisec}) fails with error {self.cam.lasterr()}')
+
+                if self.cam.buf_getframedata(-2) is False:
+                    # If there is no second to last frame, it means we just acquired the first frame.
+                    # In this case, drop the current frame since it always contains systematic errors.
+                    break
```

Comment:
While testing in the lab, we learned that the first frame of each acquisition contains very obvious systematic errors, so it is unusable. It is probably rare that anyone would use the very first frame from a data stream, but it seemed more correct to me to never submit that faulty frame to the data stream in the first place.

I am not actually sure how to test this code block though... if you have any pointers @ehpor, please let me know.

@erinpougheon could I ask you to post a screenshot of the faulty frame here in the PR, for future reference?

Reply:
Question: does this faulty frame reappear when you change ROI, exposure time, gain or other settings?

Yes, I'd never submit that frame to the data stream. I'm not sure what you mean by "test". Since this is hardware-only behaviour, you can't do unit tests unless you go for hardware-in-the-loop, which is a hassle to automate. If you mean, just testing that this behaves as expected, I'd put an info log message instead of just breaking.

Reply:
Very good questions. I'll pass them on to @erinpougheon to test in the lab when she's back on Wednesday. FYI @johanmazoyer 

I meant indeed to just test whether it behaves as expected. I'll do that.

Reply:
<img width="558" alt="Faulty frame" src="https://github.com/spacetelescope/catkit2/assets/165810820/024af14d-5571-4c48-8693-dada224cb39f">
Here is a screenshot of the faulty frame.